autoload/asyncomplete.vim

```diff
@@ -10,141 +16,87 @@ if !has('timers')
-            endif
-        endfor
-        execute 'augroup end'
+        doautocmd User asyncomplete_setup
+        let s:already_setup = 1
     endif
-
-    let s:sources[a:info['name']] = a:info
 endfunction
 
-function! asyncomplete#unregister_source(name) abort
-    try
-        call s:next_tick_single_exec('clear_active_sources', function('s:clear_active_sources'))
-        let l:info = s:sources[a:name]
-        unlet l:info
-        unlet s:sources[a:name]
-    catch
-        return
-    endtry
+function! asyncomplete#enable_for_buffer() abort
+    call s:setup_if_required()
+    let b:asyncomplete_enable = 1
 endfunction
 
-function! s:clear_active_sources() abort
-    call asyncomplete#log('core', 's:clear_active_sources', bufnr('%'))
-    if s:supports_getbufinfo
-        for l:buf in getbufinfo()
-            if has_key(l:buf['variables'], 'asyncomplete_active_sources')
-                unlet l:buf['variables']['asyncomplete_active_sources']
-            endif
-        endfor
-    endif
+function! asyncomplete#disable_for_buffer() abort
+    let b:asyncomplete_enable = 0
 endfunction
 
-function! asyncomplete#complete(name, ctx, startcol, matches, ...) abort
-    let l:refresh = a:0 > 0 ? a:1 : 0
-
-    " ignore the request if context has changed
-    if asyncomplete#context_changed(a:ctx)
-        if g:asyncomplete_force_refresh_on_context_changed
-            call s:python_cm_complete(a:name, a:ctx, a:startcol, a:matches, l:refresh, 1)
+function! asyncomplete#register_source(info) abort
+    if has_key(s:sources, a:info['name'])
+        call asyncomplete#log('core', 'duplicate asyncomplete#register_source', a:info['name'])
+        return -1
+    else
+        let s:sources[a:info['name']] = a:info
+        if has_key(a:info, 'events') && has_key(a:info, 'on_event')
+            execute 'augroup asyncomplete_source_event_' . a:info['name']
+            for l:event in a:info['events']
+                let l:exec =  'if get(b:,"asyncomplete_enable",0) | call s:notify_event_to_source("' . a:info['name'] . '", "'.l:event.'",asyncomplete#context()) | endif'
+                if type(l:event) == type('')
+                    execute 'au ' . l:event . ' * ' . l:exec
+                elseif type(l:event) == type([])
+                    execute 'au ' . join(l:event,' ') .' ' .  l:exec
+                endif
+            endfor
+            execute 'augroup end'
         endif
         return 1
     endif
-
-    call s:python_cm_complete(a:name, a:ctx, a:startcol, a:matches, l:refresh, 0)
 endfunction
 
-function! asyncomplete#force_refresh() abort
-    return asyncomplete#menu_selected() ? "\<c-y>\<c-r>=asyncomplete#_force_refresh()\<CR>" : "\<c-r>=asyncomplete#_force_refresh()\<CR>"
-endfunction
-
-function! asyncomplete#_force_refresh() abort
-    if get(b:, 'asyncomplete_enable')
-        call s:remote_refresh(asyncomplete#context(), 1)
+function! asyncomplete#unregister_source(info_or_server_name) abort
+    if type(a:info) == type({})
```

Comment:
This func dose not have argument a:info.

Reply:
Fixed.